[OG-134] projects backend

LisieckaK00 opened the pull request:
> **Note:** I'm currently opening it against **feature/OG-126-projects-frontend** so the backend changes are visible while reviewing.  
> Once MR #248 is merged, I will update this MR to target **develop**.

## SQL
There is a large SQL block that suggests variant combination, so here is a short explanation of what it does:

The query evaluates all possible combinations of variants from different categories. For example, if category X has values **A, B** and category Y has values **1, 2**, then the possible combinations are: **A1, A2, B1, B2**.

Each combination is assigned two counters:  
- `tuple_used`: how many times this exact combination (e.g. A1) has been used,  
- `sum_used`: how many times its individual category values were used (e.g, A + 1).

The final choice is the combination with the lowest `tuple_used`, and then the lowest `sum_used` ( + random at the end).

Example with existing assignments **A1** and **A2**:
`tuple_used:`
- A1 = 1  
- A2 = 1  
- B1 = 0  
- B2 = 0  

`sum_used:`  
- A1 --> **A**=2, **1**=1 --> 3  
- A2 --> **A**=2, **2**=1 --> 3  
- B1 --> **B**=0, **1**=1 --> 1  
- B2 --> **B**=0, **2**=1 --> 1  

Sorted by `(tuple_used, sum_used)`:
- B1 --> (0, 1)  
- B2 --> (0, 1)
- A1 --> (1, 3)  
- A2 --> (1, 3)  

Therefore, the least-used combinations are **B1** and **B2**, and the query returns one of them.

kligarski committed:
skeleton for project group configuration modal

kligarski committed:
context base

kligarski committed:
group pick

kligarski committed:
project variant base

kligarski committed:
add filters

kligarski committed:
accordion fix + variant picking v1

kligarski committed:
project group configuration submission

kligarski committed:
project group deletion

kligarski committed:
absolute imports

kligarski committed:
add button to randomize project variants

LisieckaK00 committed:
most get methods implemented

LisieckaK00 committed:
add suggesting variants

krudny commented:
Didn't dig deeply, but I have serious concerns of how complex is to pick least used combination, especially with time complexity of this solution.

My opinion could be absolutely wrong because i didn't read the code, but to my knowledge, we have to cross join variants with categories to get (A1, A2, B1, B2, B3) during course/project setup. And we have to do it only once. We can store results in some data structure (or db view). 

Then we could have view/data structure with courseId, combination, used_count and have query `select top 1 combination from view where course id = requested sorted by used_count asc` and taking into account our limitations, it would be nearly instant time complexity. 

I asked chat gpt to do some performance analysis, he said: 

<img width="727" height="285" alt="image" src="https://github.com/user-attachments/assets/c3630ac4-eb1f-4c97-9f2f-e7b6f3978d1f" />

<img width="726" height="187" alt="image" src="https://github.com/user-attachments/assets/30384b1f-28e2-437f-bca4-6aeae863b58d" />


which makes me more afraid than before, even if he is wrong.

if you would like to randomize variants and categories separately and somehow merge results, then ok, but i would like to be ensured, that time complexity is not 2000x higher than random heuristic and will not kill our devops. Real use case is that we have to pick 1 from 24 configurations 125 times per course per year.

To sum up, I didnt read the code, but i have serious concerns about doing iterative cross join inside recursion 125 times in short period of time (250 students / 2 people project group) and it seems as biggest overengineering in this project xD But on the other hand, you might be completely right, but i would like to see some proof and complexity analysis. I will not accept another performance issue and "someone will fix it later" conditional merge

LisieckaK00 committed:
submit + delete

krudny commented:
please update branch!!!

10/10 vulnerability, vercel refuses to deploy that

LisieckaK00 commented:
Thanks for the analysis but it’s based on an assumption that all variants in all categories participate in the cross-join. In our actual solution this is not the case, because we use a fixed **perCategoryLimit = 2**, meaning only the **two least-used variants per category** are included. That reduces the cross-join from V^C to at most 2^C combinations, which is tiny (e.g. 6 categories -> 64 combinations) and nowhere near the thousands you estimated.

Because of that, the recursive step is very small and executes comfortably within PostgreSQL's capabilities. There is **no materialization of all variant combinations**, no explosive growth, and no need for additional precomputed structures. To make the current approach run even more efficiently, I can add proper supporting indexes (on the category/variant tables), which is a lightweight and standard optimization, much simpler than building new views or caching layers.

Maintaining additional views or precomputed tables would actually introduce more complexity and operational overhead than the current on-demand approach  (that would be overengineering in my opinion).

If you still believe we should add a dedicated view or precomputed table, I’m open to updating the design, but given the real constraints the current solution is efficient and simple to maintain.

LisieckaK00 commented:
I'm aware that something isn't working correctly on the frontend when deleting a project group, but I'm not sure yet how to address it.

krudny commented:
As mentioned, the time complexity of this approach is exponential. The only reason it isn't overwhelming the infrastructure is the hardcoded perCategoryLimit = 2.

I still believe this approach is overengineered. The easiest solution would be to perform a cross join once and store variant usage statistics separately. This would make reads O(N log N) when N = 24 and updates roughly O(1), resulting in a deterministic algorithm. Alternatively, we could rely on statistics and implement a uniform random distribution.

I would also challenge you to show this 130-line SQL query to technical person knowing nothing about our app and ask them to explain what it does in a few words. I am confident that almost no one would be able to do so. To be honest, I find it difficult to review myself. The fact that you need to explain this algorithm in the PR description convinces me that it is unnecessarily complicated.

We dont have time to make drastic changes, so if you state that it works reasonably fast on dev infrastructure then okey, i will check runtime when it will be ready. But I will not review this part, because i think i am not able to.

Edit: I asked someone what it does without method name and results are obvious but I cannot paste them here as there were some uncensored words said during analysis of this SQL xD

Edit2: asked another person and reaction was even better, but due to the same reasons I cannot paste it here xD

LisieckaK00 commented:
Yes, the limit is there *exactly* to limit the computational cost (that’s the entire point of having a limit). We don’t need to evaluate all possible combinations, so we deliberately restrict the search space to keep the query lightweight.

Originally, I wanted to implement this logic in Java (which would be far more readable), but I'm confident the review feedback would then be that Java is too slow for this and we should push the computation to the database. That's exactly why the algorithm lives in SQL: because it's the fastest environment for this type of operation. Functionally it works, and locally the query returns instantly (even with the entire environment running, which already slows down the laptop). You can test this yourself.

I fully agree that the SQL is long and not ideal for code clarity. However, the statement 
> The easiest solution would be to perform a cross join once and store variant usage statistics separately. 

is not as simple operationally as it sounds. Maintaining such precomputed structures introduces new problems: updating when categories change, synchronizing on insert/delete/update, ensuring data consistency, and adding another layer of state that can drift from reality. These are non-issues today, we would be solving a problem that does not actually exist. 

As for the idea of using a uniform random distribution, it does not meet the actual requirements of the feature. A uniform random draw cannot ensure balanced variant usage across categories, nor avoid repeated full tuples, because it ignores the current state of assignments entirely. These constraints are fundamental to the problem, and a purely random distribution does not satisfy them nor provide predictable outcomes.

If it still bothers you that much, fine, I'll rewrite it into a different architecture. But in my opinion that will simply waste time without giving us any meaningful benefits beyond making the SQL shorter.

krudny commented:
<img width="517" height="302" alt="image" src="https://github.com/user-attachments/assets/3cb67925-ed44-415c-8329-b406d547f8d1" />

I am not satisfied with this, when allow cross course groups is set to true, there should be all course groups listed as it is on coordinator site. Also, when I am filtering by "wszystkie", i dont see all students.

krudny commented:
```
org.springframework.orm.jpa.JpaSystemException: Could not convert 'com.agh.polymorphia_backend.model.user.student.Animal' to 'java.lang.Long' using 'org.hibernate.type.descriptor.java.LongJavaType' to wraporg.springframework.data.jpa.repository.support.CrudMethodMetadataPostProcessor$CrudMethodMetadataPopulatingMethodInterceptor.invoke(CrudMethodMetadataPostProcessor.java:136)

2025-12-13T15:23:36.631+01:00  WARN 39693 --- [nio-8080-exec-4] .m.m.a.ExceptionHandlerExceptionResolver : Failure in @ExceptionHandler com.agh.polymorphia_backend.exception.DatabaseExceptionHandler#handleResourceNotFound(DataAccessException)

org.springframework.web.server.ResponseStatusException: 500 INTERNAL_SERVER_ERROR "Could not convert 'com.agh.polymorphia_backend.model.user.student.Animal' to 'java.lang.Long' using 
```

during project group creation. I am using my data and i think they are correct, if you would like me to use yours, then please provide reliable and consistent test data. `Example.sql` doesnt satisfy consistency and reliablity.

edit: on example sql same issue i deleted some project groups, and tried to create new ones.

krudny commented on polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/project/ProjectService.java:
```diff
@@ -72,22 +100,185 @@ public List<UserDetailsResponseDto> getAnimalProjectGroup(Long userId, Long proj
+        projectGroup.setProjectVariants(projectVariants);
+        projectGroup.setAnimals(animals);
+        projectGroupService.save(projectGroup);
+        var tmp = 3;
```
.

krudny commented on polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/project/ProjectService.java:
```diff
@@ -72,22 +100,185 @@ public List<UserDetailsResponseDto> getAnimalProjectGroup(Long userId, Long proj
+                .selectedVariants(selectedVariants)
+                .build();
+    }
+
```
transaction, maybe try catch

LisieckaK00 replied:
added

krudny commented on polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/project/ProjectGroupService.java:
```diff
@@ -24,4 +24,12 @@ public ProjectGroup findById(Long projectGroupId) {
+    public void save(ProjectGroup projectGroup) {
+        projectGroupRepository.save(projectGroup);
+    }
+
```
save and delete methods are kind of useless, they can stay as it is, but i am pointing to error handling and modifying transactions

krudny commented on polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/project/ProjectService.java:
```diff
@@ -72,22 +100,185 @@ public List<UserDetailsResponseDto> getAnimalProjectGroup(Long userId, Long proj
+    private List<Animal> getAndValidateAnimals(List<Long> studentIds, Course course, boolean allowCrossCourseGroupProjectGroups) {
+        List<Animal> animals = studentIds.stream()
+                .map(id -> animalService.getAnimal(id, course.getId())).collect(Collectors.toList());
+        if (allowCrossCourseGroupProjectGroups) {
```
i think this error is invalid when this flag is set to true

LisieckaK00 replied:
right, updated

krudny commented on polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/project/ProjectService.java:
```diff
@@ -72,22 +100,185 @@ public List<UserDetailsResponseDto> getAnimalProjectGroup(Long userId, Long proj
+    }
+
+    private List<Animal> getAndValidateAnimals(List<Long> studentIds, Course course, boolean allowCrossCourseGroupProjectGroups) {
+        List<Animal> animals = studentIds.stream()
```
do we have to serialize animals to java? i dont think so, this could be moved to this sql query in countAnimalCourseGroups using join. Also this method could be in projectRepository instead of animalService if thats more appropiate after this change. That could increase performance a bit.

LisieckaK00 replied:
But we need the `List<Animals>` anyway. This method returns the lists so then we could assign the animals to the project group (`projectGroup.setAnimals(animals)`).

krudny replied:
I am ok with List<Animals> then, but it is still invalid. I spent 10 minutes at this code and still dont get many things.

<img width="1339" height="786" alt="image" src="https://github.com/user-attachments/assets/ee6c8e87-9b9a-416a-a825-eda107f70f40" />

line 141 fetches animals inside and then do validation based on student course group assignment. I suppose that lines 158 and 166 does the same? this whole thing is messed up with what ustankie did for studnet.

And still, line 158 contains N+1 problem. For first student id we are asking our db: 

- hey, give me animal for this student id
- hi java, I will try to search, yay i found one entity 
- thank you db, i will now allocate memory and create an object and then i will continue with my super stream!

then for second student id we are asking

- hey, give me animal for this student id
- hi java, I will try to search, yay i found one entity 
- thank you db, i will now allocate memory and create an object and then i will continue with my super stream!

for 3rd time guess what, we are still asking the same question

- hey, give me animal for this student id
- hi java, I will try to search, yay i found one entity 
- thank you db, i will now allocate memory and create an object and then i will continue with my super stream!

And this can keep projectGroup.size() times. Which is not limited. I know that our usecase assumes that project group is containing 2 people. But we could ask just once regardless of how many students in project group we have:

- hey db, give me all animals for this all students ids.
- hi java, okey, i will look for this 50 animals
- thank you db, now i see that i have to allocate a lot of memory, bum, it there it is! fantastic, i can return result now.

One question. One answer. It is that easy.

Another part is: One thing is fetching animals. Second thing is validating access. Now this is mixed, no one knows why.

And most suprisingly: authorizing students could be faster, but no one is taking my todo into consideration. This is technical debt and it will be getting only worse.

Look at this part: 

<img width="966" height="124" alt="image" src="https://github.com/user-attachments/assets/6613a51c-37fd-4ad7-b888-df46e213aea4" />

You fetch Project and Course. Due to our model, you are also asking our db about event section. And that could be fine. But if you dig deeper what is invoked from this method, you will see this.
 
<img width="995" height="269" alt="image" src="https://github.com/user-attachments/assets/d211ffb9-779d-4543-a392-1b07c8ef0541" />

on the first line, you are still fetching all those things, just to have `courseId`. But lets check what is going on in access authorizer details which this method invokes.

<img width="955" height="452" alt="image" src="https://github.com/user-attachments/assets/d90d39e0-1c54-468e-9a5f-10bb86d53fa6" />

Woah, once again we fetch course. Fantastic.

The most important part. You might fetched 4 times course entity (one is from get and validate animals i talked about earlier). And guess what. You could do this 0 times. The only thing needed to authorize student is to have courseId and check student course group assignment table. And courseId could be passed in dto at the beginning from frontend. If not, then we could fetch courseId just once and this is completely fine. If we could fetch one Long variable instead of 4 times whole entity per one request it would be a superb improvement in macro scale.

ProjectService and authorizerService in my opinion have technical debt mixed with my favourite optional.ofNullable.  You can say that I am wrong in one statement, probably I am, but thats not the core point if you fetched course 3 or 4 times during one request. Its more general comment about architecture.

krudny replied:
<img width="948" height="221" alt="image" src="https://github.com/user-attachments/assets/9ffd67ab-84ab-4643-972c-3145545c3f0e" />

this method is in projectService. I have no idea for what and why there. I know that it is not your code, but just to point technical debt you all are creating there.

krudny replied:
you can create 5 more issues to my comments as ustankie did, but i suppose thats not how software engineering works

krudny replied:
Its late night, but I have to clarify more. Java streams are really useful to do things and not everything is meant to be in sql. But the problem is how you are using them without being at least aware what's going on. Ustankie in exactly every place set fetch strategy to LAZY. And sometimes yes, it has a lot of performance advantages which i like. But if you are streaming entity x and calling getter for entity y which is lazy fetched there's a classical N+1 problem. This could be avoided by being aware and not fetching things twice just because they are fetch lazy. Moreover, I strongly point that changing most things to eager fetch strategy is not a simple working solution and it can cause even worse performance of N+M+1 problem. Its kind of architectural thinking of what exactly is going on under the hood and how to use it well. 

As I see, project controller now has 9 endpoints and conceptually we have to do some crud operations on project group entity. So statistically we have endpoints to create, read, update and delete project group entity more than twice for every operation. Also, this is largest controller per one functionality and if I would guess, something might went really wrong there. 

Random comment: 

Hall of fame backend functionality has well written clean, method separated code and its really efficient. It took >3 months to finish, but it shows that you are able to write something without performance and code duplication issues.

to clarify, by saying you i meant backend engineers, not you exactly

I also feel that my too long comments are not valued in any way, as you know that in few weeks you are not going to be involved in this project so creating issue and assigning it by default to me is easier than thinking before writing some code. 

I would like to point that is 1:30am and I am not sleeping just to give some more value not only to this project, but also to you. I might click approve instead.
Personally i would love to see some more improvements than chat gpt crafted, also too long, response.

krudny commented on polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/project/ProjectService.java:
```diff
@@ -72,22 +100,185 @@ public List<UserDetailsResponseDto> getAnimalProjectGroup(Long userId, Long proj
+                        .build())
+                .collect(Collectors.toList());
+
+        long max;
```
i said somewhere, that this part is not really working and not so consistent with other parts of ui and now i can add that code quality is also poor

LisieckaK00 replied:
logic was updated

kligarski commented:
> Also, when I am filtering by "wszystkie", i dont see all students.

Note that all students that aren't assigned to any project group should be visible, not all students.

krudny commented:
>Note that all students that aren't assigned to any project group should be visible, not all students.

on my data i dont have any project group and i dont have 250 students in list as expected

krudny commented:
> 
<img width="517" height="476" alt="image" src="https://github.com/user-attachments/assets/8c39e693-5f3b-454e-984e-1bc2de83355a" />

i dont know by who, but klisiecka is assuming:

<img width="803" height="65" alt="image" src="https://github.com/user-attachments/assets/03ec143f-1e2e-4354-9ee9-58d23440025c" />

which is invalid if i could select more that 'all' option.

krudny commented on polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/project/ProjectService.java:
```diff
@@ -72,22 +100,185 @@ public List<UserDetailsResponseDto> getAnimalProjectGroup(Long userId, Long proj
+                .build();
+    }
+
+    public List<ProjectGroupPickStudentsResponseDto> getProjectGroupConfigurationGroupPickStudents(ProjectGroupPickStudentsRequestDto requestDto, Long projectId) {
```
I see another N+1 problem after isPresent() which is inefficient and this part should be in sql join. I have no idea why we have to fetch one by one hall of fame entry and student id for every animal in project group. Fetching all data in one db request is way more efficient. I dont see any limit in project group size, so if someone creates one project group with 50 targets, it will will have huge impact on performance because instead of doing all things once, we have to establish 50 db connections (network latency, creating socket connection, TCP handshake, creating a new psql process for each request, postgres query planning, execution, cleanup which is not free) just to get `studentId` from `animalId`. It is absolutely inefficient. 


This is macro scale. But suppose we have 10k users, each of them has 50 animals in project group. Numer of unnecessary queries, TCP connections, postgres processes we need to create grows exponentially. 

Random ss from random presentation about N+1 problem: 

<img width="756" height="258" alt="image" src="https://github.com/user-attachments/assets/23aaff1c-76f1-43ed-b1a8-96e203c2c2fa" />

100 ms connection is superb result. 50 connections * 100ms = 5s of waiting just to fetch 50 studentids is tragic result. It could be lowered to 100ms connection + (50 * 3ms execution) = 250ms just to get the same data. 

Yes, I know about connection pools and reusing connections, but whole argument is not about exact execution time, but about general approach to performance and code quality. 

Response looks nearly the same as student target response dto so we could use inheritance. This could be base one, for grading we could add gained xp field

krudny commented on polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/project/ProjectService.java:
```diff
@@ -72,22 +100,185 @@ public List<UserDetailsResponseDto> getAnimalProjectGroup(Long userId, Long proj
+    }
+
+    public List<ProjectGroupPickStudentsResponseDto> getProjectGroupConfigurationGroupPickStudents(ProjectGroupPickStudentsRequestDto requestDto, Long projectId) {
+        boolean includeAllGroups = requestDto.getGroups().isEmpty() ||
```
I wrote some comment that this assumption is invalid, as on frontend i can select "wszystkie" and some course groups in the same time

LisieckaK00 replied:
that was bc of misspelling, is updated now, not posisible to choose "wszystkie" and sth else

krudny commented:
public methods should be on top of file. 

<img width="1322" height="309" alt="image" src="https://github.com/user-attachments/assets/5b21c9f8-af84-4b7f-9b20-c640c8191780" />

i dont see if this method is used anywhere outside this class

krudny reviewed (changes requested):
To sum up, code quality could be better, but the most important thing is that backend could randomly throw 500 error and student part looks as semi mocked.

krudny commented:
In controller, two methods have the same name.

LisieckaK00 committed:
after review changes

kligarski committed:
fix imports after rebase

kligarski committed:
move new hooks to match new file structure

LisieckaK00 commented:
Updates:
1. Group filters: when `allow_cross_course_group_project_groups = true`, all groups are returned as filters for the instructor.
2. Selecting groups from filters: when "wszystkie" is selected, special behavior applies (thanks to @kligarski for noticing the misspelling 😆).
3. There was sometimes an error during project group creation (one of the queries was incorrect). This has now been fixed.
4. I initially assumed that course groups would be small, so I performed some mapping in Java. Updated to fetch the data directly from the database instead.

Thanks to @krudny for testing and pointing out the issues.

---

Remaining points:
1.
> You might fetched 4 times course entity

As far as I understand, this works differently. The same `Project` instance is fetched once and passed through subsequent method calls within a single request. Since the Hibernate session remains open, the first call to `project.getEventSection()` initializes the LAZY relation, and further calls do not trigger additional database queries. This is why `project.getEventSection().getCourse()` appearing multiple times in the code does not result in multiple `Course` fetches at runtime. The structure was intentional, to minimize the number of passed parameters and rely on the managed `Project` instance.

2.
> line 141 fetches animals inside and then do validation based on student course group assignment. I suppose that lines 158 and 166 does the same?

No, they do not do the same thing. 
- When editing an existing project group, I check whether the instructor has permission to edit that group. If the instructor is the owner, editing is allowed, otherwise he can't edit someone else's group (the coordinator can).
- The validation inside `getAndValidateAnimals` checks permissions for the passed animals. If `allow_cross_course_group_project_groups = true`, nothing needs to be checked. Otherwise, we need to verify that the students belong to the instructor's course group and that they are from the same group.
3. I don't understand what "student part looks as semi mocked" means.

LisieckaK00 committed:
fix and adding additional filters info

kligarski commented on polymorphia-frontend/providers/target/index.tsx:
```diff
@@ -59,6 +59,10 @@ export const TargetProvider = ({
   useEffect(() => {
+    if (isTargetsLoading) {
+      return;
+    }
```
Can you share some details why is it needed? Also, don't we need to add `isTargetsLoading` to dependency array? I did not add `selectedTarget` intentionally and that's why the warning is disabled but if we should run this code when `isTargetsLoading` changes then we should add it to dependency array.

kligarski reviewed (commented):
Should we support "Własne" group filter in grading target list for instructor as well (I see this only for coordinator) when we use cross course groups? Also, in project configuration filter configs it would be nice to return "all" as the first option instead of last.

Also, we still get error when fetching variant or group from student that is not assigned to any project group. We should return empty list instead so that the proper message is shown and tanstack query does not attempt to refetch the query.

krudny commented:
> I don't understand what "student part looks as semi mocked" means.

No one took care of it and it's still from time when it was mocked. Its not completely mocked, because data is from backend, but still this 2 pr's are not taking student view into account. Ustankie did something, but mostly her pr was rejected due to performance issues with gradable events, thats why i said it is semimocked. 

> whole first point

I gave you really robust explanation and you still didn't get correct point. Its not a matter of how many times we are asking db for course entity per one specific request. Core point is how we are using hibernate. It doesnt really matter if course entity is cached in this particular request because in another request, using access authorizer, it might be not. To me, access authorizer shouldnt fetch course entity at all. Right now you are asking 4 times for course entity and okey, its cached, but using the same pattern you can ask for 4 different entities and It would be 4N + 4 problem and none of this values will be cached in hibernate session. I am talking about APPROACH to quering database. 

As I was testing on develop db, deleting one single project group takes 1.87s. Creating project group takes more than 3s. Invalidating project groups targes after project group creation takes 10s! And all this requests are longer than fetching all gradable events with all related grades for student which can be under 1s. So thats not a hardware problem, but performance issues on code quality side. 

And due to this also too long comment i figured out that our project group entity has a foreign key to animals and it blocks cascade course group deletion so I am requesting changes, probably on db migration

LisieckaK00 committed:
update fetching project group for student, refactor accessAuthorizer

LisieckaK00 commented:
> Should we support "Własne" group filter in grading target list for instructor as well (I see this only for coordinator) when we use cross course groups?

Allowing cross course groups only affects how project groups are formed, not instructor permissions. Instructors still see only their own course groups, while the coordinator can see all groups (so for him it makes sense to have the filter). In my opinion, the current implementation makes sense.

> in project configuration filter configs it would be nice to return "all" as the first option instead of last

done

>  fetching group for student that is not assigned to any project group

I updated the logic. Now empty list is returned.

> AccessAuthorizer

I made a few changes. I refactored `AccessAuthorizer` to work only with `courseId` (instead of `Course` instances). I also updated the logic to be role-based (for example for coordinators we skip the project group membership check). This should simplify things and hopefully be a bit more efficient.

> performence

I tried to address the performance issues by moving logic to db queries where possible.

> cascade delete

not addressed yet

LisieckaK00 committed:
udpate getting projects variants for student

krudny commented:
When project has no variants, it is possible to rand combination (frontend) and it returns http 200 (backend).

krudny commented:
@kligarski, invalidating doesnt work xD I thought it was working already

i can confirm that they are created in db 

on exampledata2 also not working, i dont know how far backend is out of sync with frontend, but right now i cant accept that

https://github.com/user-attachments/assets/1513bbb8-8e1f-4791-8024-9dcd930cedd9

krudny commented on polymorphia-frontend/services/api/types.ts:
```diff
@@ -8,7 +8,7 @@ export const HttpMethods = {
 export type HttpMethod = (typeof HttpMethods)[keyof typeof HttpMethods];
 
-export type ApiBody = string | FormData | URLSearchParams | object;
+export type ApiBody = null | string | FormData | URLSearchParams | object;
```
i dont like this, increasing technical debt

Api body could be undefined if we dont pass anything and thats okey. But passing null as apibody is architectural problem, and we had a discussions with @kligarski about that and its probably our fault, but noone proposed better soltuion. Right now nullish target should be handled in hooks layers using `enabled: target != null` and using z `!` operator in function call. Alternatively early return in hook layer if target is not defined. I am struggling to live with this ! in code, but if we are having such architectural issue, then I please to make it in just one style everywhere just to not increase technical debt more. If theres null (but defined) body then we definitely shouldnt call any endpoint.

LisieckaK00 replied:
There was a different issue. Some context: null as target is allowed  in `getRandomProjectVariant` and `getProjectGroupConfigurationFilterConfigs` as these methods are used for creating new project groups (null target) and editing existing one (not null target). So backend works slightly different with and without target. I updated these methods to receive `undefined` instead of  `null`  if the target is not present

krudny commented on polymorphia-frontend/services/api/client.ts:
```diff
@@ -105,8 +105,12 @@ function put<TResponse = void>(
 
-function del(path: string, headers?: HeadersInit): Promise<void> {
-  return request({ path, method: HttpMethods.DELETE, headers });
+function del(
```
we dont use shortcuts in frontend

LisieckaK00 replied:
this was bc it's not possible to define `function delete` in ts. To avoid the shortcut I moved the definitions inside ApiClient directly

krudny commented on polymorphia-frontend/hooks/course/projects/useProjectGroupDelete/index.tsx:
```diff
@@ -21,24 +22,49 @@ export default function useProjectGroupDelete(): UseProjectGroupDelete {
       queryClient.invalidateQueries({
         queryKey: ["gradingTargets"],
       });
       queryClient.invalidateQueries({
```
this solution is really bad, to be honest i dont fully understand why we are using `refetchType: none` there.

1. Suppose we should use it:

Then you can create array of query keys and do:

```ts
queriesToInvalidate.forEach((queryKey) => {
    queryClient.invalidateQueries({
      queryKey,
      refetchType: "none",
    });
  });
```

less code, much cleaner 

2. case if we shouldnt do that. Docs are saying that `refetchType: none` makes this data stale without refetching. If the purpose of this code is that we dont want to invalidate queries for not existing keys then we can simply remove queryKeys from tanstack cache using:

```ts
onSuccess: (_, { target }) => {
  queryClient.invalidateQueries({ queryKey: ["gradingTargets"] });
  queryClient.removeQueries({
    predicate: (query) => 
  });
```

but i have serious doubts why we do such things only here, maybe its llm trash

I am just guessing, but if there is a problem that you delete target but target is still selected in target context and tanstack query is refetching data for not existing entity, then you should remove selected target from context after successful deletion, but i suppose that after invalidating `gradingTargets` selected target should also be changed to new one. 

I cant test, because grading targets are not refetched xD

LisieckaK00 replied:
It's funny bc I had a lot of issues the target lists and that was the one solution that was working (I tried sth with removing the queries, don't remember anymore what excatly). But now it works without this code entirely XD I guess the update on target lists fixed whatever issue I had here. So code removed

krudny commented:
Projects in exampledata2 are hidden, please change

krudny commented:
@kligarski student side

<img width="622" height="909" alt="image" src="https://github.com/user-attachments/assets/3ccf166c-9556-4b73-9569-305a78328a43" />

krudny commented on polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/validation/AccessAuthorizer.java:
```diff
@@ -8,6 +8,7 @@
+import com.agh.polymorphia_backend.repository.project.ProjectGroupRepository;
 import com.agh.polymorphia_backend.repository.user.UserCourseRoleRepository;
 import com.agh.polymorphia_backend.repository.user.role.InstructorRepository;
 import com.agh.polymorphia_backend.repository.user.role.StudentRepository;
```
there's todo to remove, as problem is seems to be solved

LisieckaK00 replied:
removed

krudny commented on polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/validation/AccessAuthorizer.java:
```diff
@@ -33,61 +34,45 @@ public class AccessAuthorizer {
     private final CourseRepository courseRepository;
+    private final ProjectGroupRepository projectGroupRepository;
 
     public void authorizeCourseAccess(Long courseId) {
```
maybe authorizeCurrentUserCourseAccess?

LisieckaK00 replied:
updated

krudny commented on polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/validation/AccessAuthorizer.java:
```diff
@@ -111,47 +96,49 @@ public boolean hasAnyRole(List<UserType> roles) {
 
-    private boolean hasInstructorAccessToUserInCourse(User user, Course course, Long studentId) {
-        return instructorRepository.hasAccessToStudentInCourse(user.getId(), course.getId(), studentId);
+    private boolean hasInstructorAccessToUserInCourse(Long userId, Long courseId, Long studentId) {
```
i think "toStudent"

LisieckaK00 replied:
deleted the method

krudny commented:
<img width="728" height="254" alt="image" src="https://github.com/user-attachments/assets/665ac2d7-152e-48ea-a119-532b419d20db" />

is this `type(tru) = Instructor` needed? if teaching role is coordinator, then whole result will be false, but should be true by default. We are ensuring in access authorizer that we are asking about instructor role, but every coordinator has higher access rules than instructor sooo no point to check that? i dont know

krudny commented on polymorphia-backend/src/main/java/com/agh/polymorphia_backend/repository/course/CourseRepository.java:
```diff
@@ -2,8 +2,15 @@
 
 @Repository
 public interface CourseRepository extends JpaRepository<Course, Long> {
+
```
why not comparing in sql? would be faster 

```java
@Query("""
    select count(c) > 0
    from Course c
    where c.id = :courseId and c.coordinator.userId = :userId
""")
boolean isCourseCoordinatorAuthorized(Long userId, Long courseId);
```

LisieckaK00 replied:
updated

krudny commented on polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/validation/AccessAuthorizer.java:
```diff
@@ -111,47 +96,49 @@ public boolean hasAnyRole(List<UserType> roles) {
-    private boolean isCourseAccessAuthorizedInstructor(User user, Course course) {
-        return instructorRepository.findByUserIdAndCourseId(user.getId(), course.getId()).isPresent();
+    private boolean isCourseAccessAuthorizedInstructor(Long userId, Long courseId) {
+        return instructorRepository.findByUserIdAndCourseId(userId, courseId).isPresent();
```
i know it is not super important, but if we are refactoring to be more performant then we could value small things and do it properly. Fetching whole instructor entity just to check it exists and return boolean is useless. `boolean existsByUserIdAndCourseId(Long userId, Long courseId);` would be much faster than fetching entity and wrapping with java optional.

LisieckaK00 replied:
you're right, I didn't notice that. updated now

krudny commented on polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/validation/AccessAuthorizer.java:
```diff
@@ -33,61 +34,45 @@ public class AccessAuthorizer {
+            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Brak dostępu do grupy projektowej.");
         }
     }
 
```
Why this method below accepts course not courseId? in method invocation we have `Course course = courseService.getCourseById(courseId);`, so it is useless and it should be refactored to accept courseId. Same case with animal, its useless to fetch whole animal entity just to confirm it exists, please use `boolean existsByUserIdAndCourseId(Long userId, Long courseId);`

LisieckaK00 replied:
updated, but in the `UserContextService::setPreferredCourseId` the course entity is fetched anyway bc:
```java
 dbUser.setPreferredCourse(courseService.getCourseById(courseId));
```

krudny commented on polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/knowledge_base/KnowledgeBaseService.java:
```diff
@@ -56,9 +55,7 @@ private <T> List<KnowledgeBaseResponseDto> getKnowledgeBaseResponseDto(
             Function<Long, List<T>> repositoryFetcher,
             Function<T, KnowledgeBaseResponseDto> mapper
     ) {
-        Course course = courseService.getCourseById(courseId);
```
thats what i like, perfect

krudny commented on polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/gradable_event/GradableEventService.java:
```diff
@@ -70,7 +71,7 @@ public List<BaseGradableEventResponseDto> getGradableEvents(Long relatedId, Grad
             case EVENT_SECTION -> courseService.getCourseByEventSectionId(relatedId);
```
i suppose that there you can use .getId() and below pass just courseId everywhere instead of course entity

LisieckaK00 replied:
updated

krudny commented on polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/gradable_event/GradableEventService.java:
```diff
@@ -83,6 +84,11 @@ public List<BaseGradableEventResponseDto> getGradableEvents(Long relatedId, Grad
         };
     }
 
+    public Long getCourseIdByGradableEventId(Long gradableEventId) {
```
i think that `fetchGradableEvent` method in this file is useless and could be inlined were it is used. I suppose that previously it had more usages, thats why it was separated by me, right now it has one usage and private scope.

LisieckaK00 replied:
removed

kligarski commented:
> @kligarski student side
> <img alt="image" width="622" height="909" src="https://private-user-images.githubusercontent.com/134285490/530237699-3ccf166c-9556-4b73-9569-305a78328a43.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjY4MzEyNzksIm5iZiI6MTc2NjgzMDk3OSwicGF0aCI6Ii8xMzQyODU0OTAvNTMwMjM3Njk5LTNjY2YxNjZjLTk1NTYtNGI3My05NTY5LTMwNWE3ODMyOGE0My5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMjI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTIyN1QxMDIyNTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hYTU1MGMwYmQ2MzQzMjVkYWM0M2M3MGRlMjc4Y2M3Y2IwZjdjODY3ZjA0ZjViZDIyZjEyOTRmOTk4YzBkMTZhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.N8nEe5iJLt_1gRxCvPUiVv60N3FUdZ_D8O9jvSD-6Ng">

https://github.com/krudny/Polymorphia/pull/248/commits/61c8ab195e41a95cefc6f81573010a4385f8496e

LisieckaK00 committed:
some after review changes

LisieckaK00 commented:
for now just backend updates (removing project groups on course group delete is not implemented yet)

LisieckaK00 committed:
mostly frontend updated

LisieckaK00 committed:
delete project group on course group removal

LisieckaK00 commented:
I think I addressed all issues @krudny. Fixed example-data (to have criteria for project). Disbaled buton to rand combination when there are none varaint categories. Implemented removing project groups when deleting course group.

LisieckaK00 committed:
fix equipment sections titles

LisieckaK00 committed:
Merge remote-tracking branch 'origin/develop' into feature/OG-134-projects-backend
